agent/arbiter-agent.c

```diff
@@ -198,19 +198,90 @@ chamge_arbiter_agent_class_init (ChamgeArbiterAgentClass * klass)
 }
 
 static gboolean
-chamge_arbiter_agent_handle_enroll (ChamgeDBusArbiterManager * manager,
-    GDBusMethodInvocation * invocation, gpointer user_data)
+chamge_arbiter_agent_enroll (ChamgeArbiterAgent * self)
 {
-  ChamgeArbiterAgent *self = (ChamgeArbiterAgent *) user_data;
   ChamgeReturn ret = CHAMGE_RETURN_FAIL;
+  ChamgeNodeState state = CHAMGE_NODE_STATE_NULL;
 
   if (self->arbiter == NULL)
     g_error ("arbiter is NULL");
+
+  /* check state to prevent terminating agent
+   * in case of it is called redundantly by dbus */
+  g_object_get (CHAMGE_NODE (self->arbiter), "state", &state, NULL);
+  if (state != CHAMGE_NODE_STATE_NULL) {
+    g_debug ("node state is not CHAMGE_NODE_STATE_NULL");
+    goto out;
+  }
+
   ret = chamge_node_enroll (CHAMGE_NODE (self->arbiter), FALSE);
-  if (ret != CHAMGE_RETURN_OK) {
+  if (ret != CHAMGE_RETURN_OK)
     g_error ("arbiter enroll failure");
+
+out:
+  return ret;
+}
+
+static gboolean
+chamge_arbiter_agent_activate (ChamgeArbiterAgent * self)
+{
+  ChamgeReturn ret = CHAMGE_RETURN_FAIL;
+  ChamgeNodeState state = CHAMGE_NODE_STATE_NULL;
+
+  if (self->arbiter == NULL)
+    g_error ("arbiter is NULL");
```

Comment:
It's not a big deal, but it's simple to use assertion. `g_return_val_if_fail (CHAMGE_IS_ARBITER_AGENT (self), FALSE)`